In [1]:
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_digits

In [2]:
digits=load_digits()
dir(digits)

['DESCR', 'data', 'images', 'target', 'target_names']

In [3]:
digits.data

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [4]:
digits.target

array([0, 1, 2, ..., 8, 9, 8])

In [5]:
digits.images

array([[[ 0.,  0.,  5., ...,  1.,  0.,  0.],
        [ 0.,  0., 13., ..., 15.,  5.,  0.],
        [ 0.,  3., 15., ..., 11.,  8.,  0.],
        ...,
        [ 0.,  4., 11., ..., 12.,  7.,  0.],
        [ 0.,  2., 14., ..., 12.,  0.,  0.],
        [ 0.,  0.,  6., ...,  0.,  0.,  0.]],

       [[ 0.,  0.,  0., ...,  5.,  0.,  0.],
        [ 0.,  0.,  0., ...,  9.,  0.,  0.],
        [ 0.,  0.,  3., ...,  6.,  0.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.]],

       [[ 0.,  0.,  0., ..., 12.,  0.,  0.],
        [ 0.,  0.,  3., ..., 14.,  0.,  0.],
        [ 0.,  0.,  8., ..., 16.,  0.,  0.],
        ...,
        [ 0.,  9., 16., ...,  0.,  0.,  0.],
        [ 0.,  3., 13., ..., 11.,  5.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.]],

       ...,

       [[ 0.,  0.,  1., ...,  1.,  0.,  0.],
        [ 0.,  0., 13., ...,  2.,  1.,  0.],
        [ 0.,  0., 16., ..., 16.,  5.,  0.

In [6]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [7]:
import pandas as pd
df=pd.DataFrame(digits.data,digits.target)
df.head()

0    1    2     3     4     5    6    7    8    9   ...   54   55   56  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    57   58    59    60    61   62   63  
0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4  0.0  0.0   2.0  16.0   4.0  0.0  0.0  

[5 rows x 64 columns]

In [8]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest=train_test_split(digits.data,digits.target,test_size=0.3)

In [9]:
len(xtrain)

1257

In [10]:
from sklearn.model_selection import cross_val_score
cross_val_score(RandomForestClassifier(criterion='entropy',n_estimators=10),digits.data,digits.target,cv=5)

array([0.90555556, 0.86111111, 0.92200557, 0.95543175, 0.8913649 ])

In [11]:
cross_val_score(RandomForestClassifier(criterion='gini',n_estimators=10),digits.data,digits.target,cv=5)

array([0.90277778, 0.85833333, 0.93314763, 0.90807799, 0.86908078])

In [12]:
cross_val_score(LogisticRegression(solver='newton-cg'),digits.data,digits.target,cv=5)

array([0.925     , 0.87777778, 0.93871866, 0.93314763, 0.89693593])

In [13]:
cross_val_score(LogisticRegression(solver='liblinear'),digits.data,digits.target,cv=5)

array([0.92222222, 0.88333333, 0.95264624, 0.95821727, 0.89415042])

In [14]:
cross_val_score(GaussianNB(),digits.data,digits.target,cv=5)

array([0.78055556, 0.78333333, 0.79387187, 0.8718663 , 0.80501393])

In [15]:
cross_val_score(MultinomialNB(),digits.data,digits.target,cv=5)

array([0.88611111, 0.83611111, 0.8551532 , 0.94150418, 0.83286908])

In [16]:
cross_val_score(DecisionTreeClassifier(),digits.data,digits.target,cv=5)

array([0.78055556, 0.69444444, 0.7994429 , 0.82172702, 0.79387187])

In [17]:
cross_val_score(DecisionTreeClassifier(criterion='entropy',splitter='random'),digits.data,digits.target,cv=5)

array([0.79166667, 0.73611111, 0.80779944, 0.84958217, 0.76601671])

In [18]:
cross_val_score(svm.SVC(C=5,kernel='linear'),digits.data,digits.target,cv=5)

array([0.96388889, 0.91944444, 0.96657382, 0.9637883 , 0.92479109])

In [19]:
# for SVC
import numpy as np
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(svm.SVC(gamma='auto'), {
    'C': np.arange(1,10),
    'kernel': ['rbf','linear','poly']
}, cv=5, return_train_score=False)
clf.fit(digits.data, digits.target)
clf.cv_results_

{'mean_fit_time': array([0.35369658, 0.02898741, 0.03633637, 0.35804043, 0.02988892,
        0.0404253 , 0.34854846, 0.03185973, 0.03524098, 0.35048795,
        0.03191128, 0.03454247, 0.89623208, 0.09862232, 0.09871764,
        0.99001369, 0.0816565 , 0.10854697, 1.05836372, 0.08883109,
        0.0938139 , 1.06679835, 0.09341583, 0.10726252, 0.96809993,
        0.07968678, 0.13393197]),
 'std_fit_time': array([0.01082101, 0.0055975 , 0.00449857, 0.01453781, 0.00308479,
        0.00328906, 0.01005426, 0.00285083, 0.00325791, 0.00842936,
        0.00280696, 0.003886  , 0.19615404, 0.01332716, 0.00598291,
        0.02741801, 0.00488311, 0.01502517, 0.06644186, 0.01173974,
        0.00640537, 0.14100762, 0.01040728, 0.01140196, 0.07070606,
        0.0095296 , 0.01550523]),
 'mean_score_time': array([0.04056849, 0.00859275, 0.01032534, 0.04186759, 0.00807581,
        0.00727377, 0.04208455, 0.00800996, 0.00991163, 0.0415535 ,
        0.00762234, 0.01161075, 0.11320477, 0.02425027, 0.028048

In [20]:
newdf=pd.DataFrame(clf.cv_results_)
newdf.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0       0.353697      0.010821         0.040568        0.003645       1   
1       0.028987      0.005598         0.008593        0.001490       1   
2       0.036336      0.004499         0.010325        0.002504       1   
3       0.358040      0.014538         0.041868        0.003088       2   
4       0.029889      0.003085         0.008076        0.000112       2   

  param_kernel                        params  split0_test_score  \
0          rbf     {'C': 1, 'kernel': 'rbf'}           0.411111   
1       linear  {'C': 1, 'kernel': 'linear'}           0.963889   
2         poly    {'C': 1, 'kernel': 'poly'}           0.983333   
3          rbf     {'C': 2, 'kernel': 'rbf'}           0.452778   
4       linear  {'C': 2, 'kernel': 'linear'}           0.963889   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.450000           0.454039           0.448468           0.479109   
1           0.919444           0.966574           0.963788           0.924791   
2           0.944444           0.980501           0.988858           0.947075   
3           0.469444           0.479109           0.479109           0.501393   
4           0.919444           0.966574           0.963788           0.924791   

   mean_test_score  std_test_score  rank_test_score  
0         0.448545        0.021761               27  
1         0.947697        0.020978               10  
2         0.968842        0.019056                1  
3         0.476366        0.015784               19  
4         0.947697        0.020978               10

In [21]:
newdf[['params','mean_test_score']]

params  mean_test_score
0      {'C': 1, 'kernel': 'rbf'}         0.448545
1   {'C': 1, 'kernel': 'linear'}         0.947697
2     {'C': 1, 'kernel': 'poly'}         0.968842
3      {'C': 2, 'kernel': 'rbf'}         0.476366
4   {'C': 2, 'kernel': 'linear'}         0.947697
5     {'C': 2, 'kernel': 'poly'}         0.968842
6      {'C': 3, 'kernel': 'rbf'}         0.476366
7   {'C': 3, 'kernel': 'linear'}         0.947697
8     {'C': 3, 'kernel': 'poly'}         0.968842
9      {'C': 4, 'kernel': 'rbf'}         0.476366
10  {'C': 4, 'kernel': 'linear'}         0.947697
11    {'C': 4, 'kernel': 'poly'}         0.968842
12     {'C': 5, 'kernel': 'rbf'}         0.476366
13  {'C': 5, 'kernel': 'linear'}         0.947697
14    {'C': 5, 'kernel': 'poly'}         0.968842
15     {'C': 6, 'kernel': 'rbf'}         0.476366
16  {'C': 6, 'kernel': 'linear'}         0.947697
17    {'C': 6, 'kernel': 'poly'}         0.968842
18     {'C': 7, 'kernel': 'rbf'}         0.476366
19  {'C': 7, 'kernel': 'linear'}         0.947697
20    {'C': 7, 'kernel': 'poly'}         0.968842
21     {'C': 8, 'kernel': 'rbf'}         0.476366
22  {'C': 8, 'kernel': 'linear'}         0.947697
23    {'C': 8, 'kernel': 'poly'}         0.968842
24     {'C': 9, 'kernel': 'rbf'}         0.476366
25  {'C': 9, 'kernel': 'linear'}         0.947697
26    {'C': 9, 'kernel': 'poly'}         0.968842

In [22]:
clf.best_estimator_

SVC(C=1, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [23]:
clf.best_params_

{'C': 1, 'kernel': 'poly'}

In [24]:
clf.best_score_

0.96884246363355

In [26]:
model_params = {
    'svm': {
        'model': svm.SVC(gamma='auto'),
        'params' : {
            'C': [1,10,20],
            'kernel': ['rbf','linear']
        }  
    },
    'random_forest': {
        'model': RandomForestClassifier(),
        'params' : {
            'n_estimators': [1,10,20]
        }
    },
    'logistic_regression' : {
        'model': LogisticRegression(solver='liblinear',multi_class='auto'),
        'params': {
            'C': [1,10,20]
        }
    },
    'GausianNB':{
        'model':GaussianNB(),
        'params':{}
        
    },
    'MultinomialNB':{
        'model':MultinomialNB(),
        'params':{}
    },
    'decisiontree':{
        'model':DecisionTreeClassifier(),
        'params':{
            'criterion':['gini','entropy'],
            'splitter':['best','random']
        }
    }
}

In [32]:
scores = []

for model_name, mp in model_params.items():
    clf =  GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf.fit(digits.data, digits.target)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })


In [33]:
scoredf=pd.DataFrame(scores,columns=['model','best_score','best_params'])
scoredf

model  best_score  \
0                  svm    0.947697   
1        random_forest    0.928801   
2  logistic_regression    0.922114   
3            GausianNB    0.806928   
4        MultinomialNB    0.870350   
5         decisiontree    0.814704   

                                      best_params  
0                    {'C': 1, 'kernel': 'linear'}  
1                            {'n_estimators': 20}  
2                                        {'C': 1}  
3                                              {}  
4                                              {}  
5  {'criterion': 'entropy', 'splitter': 'random'}

###### So the best classifier be SVM:-	svm	0.947697	{'C': 1, 'kernel': 'linear'}